### PR TITLE
Request system initial1.10

### PR DESCRIPTION
**Pull request by Raycoms**
Ports and merges @OrionDevelopment request system to 1.10

Done by @Asherslab 

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/locations/EntityLocation.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+     * @return True when reachable, false when not.
+     */
+    @Override
+    public boolean isReachableFromLocation(@NotNull final ILocation location)
+    {
+        checkEntity();
+        return !(entity == null || entity.get() == null) && location.getDimension() == getDimension();
+    }
+
+    @SuppressWarnings("squid:S2972")
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/locations/EntityLocation.java:97]: INFO Suppressing the 'squid:S2972' warning is not allowed [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1309)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/locations/EntityLocation.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+         * Moving the curly braces really makes the code hard to read.
+         */
+        public TypeToken<EntityLocation> getFactoryOutputType()
+        {
+            return new TypeToken<EntityLocation>() {};
+        }
+
+        @NotNull
+        @Override
+        @SuppressWarnings("squid:LeftCurlyBraceStartLineCheck")
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/locations/EntityLocation.java:122]: INFO Suppressing the 'squid:LeftCurlyBraceStartLineCheck' warning is not allowed [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1309)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/locations/EntityLocation.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    {
+
+        ////// --------------------------- NBTConstants --------------------------- \\\\\\
+        private static final String NBT_MSB = "Id_MSB";
+        private static final String NBT_LSB = "Id_LSB";
+        ////// --------------------------- NBTConstants --------------------------- \\\\\\
+
+        @NotNull
+        @Override
+        @SuppressWarnings("squid:LeftCurlyBraceStartLineCheck")
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/locations/EntityLocation.java:111]: INFO Suppressing the 'squid:LeftCurlyBraceStartLineCheck' warning is not allowed [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1309)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseRequestResolver.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    {
+        //TODO Release the stack from the warehouse.
+        return null;
+    }
+
+    @Nullable
+    @Override
+    public void onResolvingOverruled(@NotNull final IRequestManager manager, @NotNull final IRequest<ItemStack> request)
+    {
+        //TODO Release the stack from the warehouse.
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseRequestResolver.java:116]: INFO Complete the task associated to this TODO comment. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseRequestResolver.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    @SuppressWarnings("squid:S2972")
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseRequestResolver.java:119]: INFO Suppressing the 'squid:S2972' warning is not allowed [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1309)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseRequestResolver.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        {
+            warehouseRequestResolver.onRequestComplete(itemStackRequestToken);
+        }
+    }
+
+    @NotNull
+    @Override
+    public void onRequestComplete(@NotNull final IToken token)
+    {
+        //TODO Release the DMan that did the job.
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseRequestResolver.java:167]: INFO Complete the task associated to this TODO comment. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseRequestResolver.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    {
+        //No followup needed.
+        return null;
+    }
+
+    @Nullable
+    @Override
+    public IRequest onParentCancelled(@NotNull final IRequestManager manager, @NotNull final IRequest<ItemStack> request)
+    {
+        //TODO Release the stack from the warehouse.
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseRequestResolver.java:108]: INFO Complete the task associated to this TODO comment. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseRequestResolver.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+            return wareHouse.isInHut(requestToCheck.getRequest());
+        }
+
+        return false;
+    }
+
+    @Nullable
+    @Override
+    @SuppressWarnings("squid:LeftCurlyBraceStartLineCheck")
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseRequestResolver.java:63]: INFO Suppressing the 'squid:LeftCurlyBraceStartLineCheck' warning is not allowed [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1309)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/locations/StaticLocation.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    {
+
+        ////// --------------------------- NBTConstants --------------------------- \\\\\\
+        private static final String NBT_POS = "Pos";
+        private static final String NBT_DIM = "Dim";
+        ////// --------------------------- NBTConstants --------------------------- \\\\\\
+
+        @NotNull
+        @Override
+        @SuppressWarnings("squid:LeftCurlyBraceStartLineCheck")
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/locations/StaticLocation.java:82]: INFO Suppressing the 'squid:LeftCurlyBraceStartLineCheck' warning is not allowed [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1309)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/locations/StaticLocation.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    @Override
+    public boolean isReachableFromLocation(@NotNull final ILocation location)
+    {
+        return location.getDimension() == getDimension();
+    }
+
+    /**
+     * Internal factory class.
+     */
+    @SuppressWarnings("squid:S2972")
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/locations/StaticLocation.java:68]: INFO Suppressing the 'squid:S2972' warning is not allowed [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1309)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/locations/StaticLocation.java [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+         * Moving the curly braces really makes the code hard to read.
+         */
+        public TypeToken<StaticLocation> getFactoryOutputType()
+        {
+            return new TypeToken<StaticLocation>() {};
+        }
+
+        @NotNull
+        @Override
+        @SuppressWarnings("squid:LeftCurlyBraceStartLineCheck")
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/locations/StaticLocation.java:93]: INFO Suppressing the 'squid:LeftCurlyBraceStartLineCheck' warning is not allowed [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1309)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+
+            final List<IToken> childTokens = new ArrayList<>();
+            final NBTTagList childCompound = nbt.getTagList(NBT_CHILDREN, Constants.NBT.TAG_COMPOUND);
+            for (int i = 0; i < childCompound.tagCount(); i++)
+            {
+                childTokens.add(controller.deserialize(childCompound.getCompoundTagAt(i)));
+            }
+
+            @SuppressWarnings(Suppression.LEFT_CURLY_BRACE)
+            final StandardRequests.ItemStackRequest request = controller.getNewInstance(requested, new TypeToken<StandardRequests.ItemStackRequest>() {}, token, state);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java:147]: MINOR Move this left curly brace to the beginning of next line of code. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+            {
+                request.setResult(ItemStack.loadItemStackFromNBT(nbt.getCompoundTag(NBT_RESULT)));
+            }
+
+            return request;
+        }
+    }
+
+    @SuppressWarnings(Suppression.BIG_CLASS)
+    public static final class DeliveryFactory implements IRequestFactory<Delivery, StandardRequests.DeliveryRequest>
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java:164]: MAJOR Reduce this class from 80 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+        {
+            return new TypeToken<StandardRequests.DeliveryRequest>() {};
+        }
+
+        @NotNull
+        @Override
+        @SuppressWarnings(Suppression.LEFT_CURLY_BRACE)
+        public TypeToken<Delivery> getFactoryInputType()
+        {
+            return new TypeToken<Delivery>() {};
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java:180]: MINOR Move this left curly brace to the beginning of next line of code. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+        {
+            return new StandardRequests.ItemStackRequest(location, token, initialState, input);
+        }
+
+        @NotNull
+        @Override
+        @SuppressWarnings(Suppression.LEFT_CURLY_BRACE)
+        public TypeToken<StandardRequests.ItemStackRequest> getFactoryOutputType()
+        {
+            return new TypeToken<StandardRequests.ItemStackRequest>() {};
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java:70]: MINOR Move this left curly brace to the beginning of next line of code. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+
+            final List<IToken> childTokens = new ArrayList<>();
+            final NBTTagList childCompound = nbt.getTagList(NBT_CHILDREN, Constants.NBT.TAG_COMPOUND);
+            for (int i = 0; i < childCompound.tagCount(); i++)
+            {
+                childTokens.add(controller.deserialize(childCompound.getCompoundTagAt(i)));
+            }
+
+            @SuppressWarnings(Suppression.LEFT_CURLY_BRACE)
+            final StandardRequests.DeliveryRequest request = controller.getNewInstance(requested, new TypeToken<StandardRequests.DeliveryRequest>() {}, token, state);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java:249]: MINOR Move this left curly brace to the beginning of next line of code. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+
+    /**
+     * Private constructor to hide the implicit public one.
+     */
+    private StandardRequestFactories()
+    {
+    }
+
+    @SuppressWarnings(Suppression.BIG_CLASS)
+    public static final class ItemStackFactory implements IRequestFactory<ItemStack, StandardRequests.ItemStackRequest>
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java:44]: MAJOR Reduce this class from 80 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java [hunk 18]
```diff
@@ -0,0 +1,283 @@
...
+    @SuppressWarnings(Suppression.BIG_CLASS)
+    public static final class DeliveryFactory implements IRequestFactory<Delivery, StandardRequests.DeliveryRequest>
+    {
+
+        @NotNull
+        @Override
+        @SuppressWarnings(Suppression.LEFT_CURLY_BRACE)
+        public TypeToken<StandardRequests.DeliveryRequest> getFactoryOutputType()
+        {
+            return new TypeToken<StandardRequests.DeliveryRequest>() {};
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java:172]: MINOR Move this left curly brace to the beginning of next line of code. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java [hunk 19]
```diff
@@ -0,0 +1,283 @@
...
+        {
+            return new TypeToken<StandardRequests.ItemStackRequest>() {};
+        }
+
+        @NotNull
+        @Override
+        @SuppressWarnings(Suppression.LEFT_CURLY_BRACE)
+        public TypeToken<ItemStack> getFactoryInputType()
+        {
+            return new TypeToken<ItemStack>() {};
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequestFactories.java:78]: MINOR Move this left curly brace to the beginning of next line of code. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/AbstractRequest.java [hunk 20]
```diff
@@ -0,0 +1,407 @@
...
+    private final List<IToken> children;
+    @NotNull
+    private final IRequester   requester;
+    @NotNull
+    private RequestState state = RequestState.CREATED;
+    @Nullable
+    private R      result;
+    @Nullable
+    private IToken parent;
+    @SuppressWarnings("squid:S1170")
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/AbstractRequest.java:42]: INFO Suppressing the 'squid:S1170' warning is not allowed [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1309)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 81 issues
* ![MAJOR][MAJOR] 24 major
* ![MINOR][MINOR] 43 minor
* ![INFO][INFO] 14 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Colony.java](https://github.com/Minecolonies/minecolonies/blob/370bc8d0a5c946f06015aac42a33e22d3bdbfc6c/src/main/java/com/minecolonies/coremod/colony/Colony.java): This file has 1.302 lines, which is greater than 1.000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [Colony.java#L53](https://github.com/Minecolonies/minecolonies/blob/370bc8d0a5c946f06015aac42a33e22d3bdbfc6c/src/main/java/com/minecolonies/coremod/colony/Colony.java#L53): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 45 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [Colony.java#L53](https://github.com/Minecolonies/minecolonies/blob/370bc8d0a5c946f06015aac42a33e22d3bdbfc6c/src/main/java/com/minecolonies/coremod/colony/Colony.java#L53): class "Colony" has 88 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [Colony.java#L680](https://github.com/Minecolonies/minecolonies/blob/370bc8d0a5c946f06015aac42a33e22d3bdbfc6c/src/main/java/com/minecolonies/coremod/colony/Colony.java#L680): Reduce the number of returns of this method 11, down to the maximum allowed 6. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
1. ![MAJOR][MAJOR] [Colony.java#L873](https://github.com/Minecolonies/minecolonies/blob/370bc8d0a5c946f06015aac42a33e22d3bdbfc6c/src/main/java/com/minecolonies/coremod/colony/Colony.java#L873): Refactor this method to reduce its Cognitive Complexity from 22 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [Colony.java#L1195](https://github.com/Minecolonies/minecolonies/blob/370bc8d0a5c946f06015aac42a33e22d3bdbfc6c/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1195): The Cyclomatic Complexity of this method "onWorldTick" is 17 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [Colony.java#L1350](https://github.com/Minecolonies/minecolonies/blob/370bc8d0a5c946f06015aac42a33e22d3bdbfc6c/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1350): Assign this magic number 16 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [Colony.java#L1352](https://github.com/Minecolonies/minecolonies/blob/370bc8d0a5c946f06015aac42a33e22d3bdbfc6c/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1352): Assign this magic number 16 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [Colony.java#L1354](https://github.com/Minecolonies/minecolonies/blob/370bc8d0a5c946f06015aac42a33e22d3bdbfc6c/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1354): Assign this magic number 128 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [Colony.java#L1473](https://github.com/Minecolonies/minecolonies/blob/370bc8d0a5c946f06015aac42a33e22d3bdbfc6c/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1473): Assign this magic number 0.5D to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by marchermans — APPROVED** [review]
From what i have implemented this looks good.

**Pull request merged**